pkg/splunk/enterprise/util.go

```diff
@@ -34,15 +37,244 @@ import (
+}
+
+// GetNamespaceScopedSecret retreives namespace scoped secret
+func GetNamespaceScopedSecret(c splcommon.ControllerClient, namespace string) (*corev1.Secret, error) {
+	var namespaceScopedSecret corev1.Secret
+
+	// Check if a namespace scoped secret exists
+	namespacedName := types.NamespacedName{Namespace: namespace, Name: GetNamespaceScopedSecretName(namespace)}
+	err := c.Get(context.TODO(), namespacedName, &namespaceScopedSecret)
+	if err != nil {
+		// Didn't find it
+		return nil, err
+	}
+
+	return &namespaceScopedSecret, nil
+}
+
+// GetVersionedSecretVersion checks if the secretName includes the versionedSecretIdentifier and if so, extracts the version
+func GetVersionedSecretVersion(secretName string, versionedSecretIdentifier string) (int, error) {
+	// Extracting version from secret's name
+	version := strings.TrimPrefix(secretName, GetVersionedSecretName(versionedSecretIdentifier, ""))
+
+	// Check if the secretName includes the versionedSecretIdentifier
+	if version != secretName {
+		// Includes, version extracted, check if version number is valid
+		versionInt, err := strconv.Atoi(version)
+		if err != nil {
+			return -1, errors.New(nonIntegerVersionError)
+		}
+
+		// Versions should be > 0
+		if versionInt <= 0 {
+			return -1, errors.New(lessThanOrEqualToZeroVersionError)
+		}
+
+		return versionInt, nil
+	}
+
+	// Secret name not matching required criteria
+	return -1, fmt.Errorf(nonMatchingStringError, secretName, versionedSecretIdentifier)
+}
+
+// GetExistingLatestVersionedSecret retreives latest EXISTING versionedSecretIdentifier based secret existing currently in the namespace
+func GetExistingLatestVersionedSecret(c splcommon.ControllerClient, namespace string, versionedSecretIdentifier string) (*corev1.Secret, int, bool) {
+	scopedLog := log.WithName("GetExistingLatestVersionedSecret").WithValues(
+		"versionedSecretIdentifier", versionedSecretIdentifier,
+		"namespace", namespace)
+
+	// Get list of secrets in K8S cluster
+	secretList := corev1.SecretList{}
+
+	// Retreive only secrets only from namespace
+	listOpts := []client.ListOption{
+		client.InNamespace(namespace),
+	}
+
+	err := c.List(context.TODO(), &secretList, listOpts...)
+	if err != nil || len(secretList.Items) == 0 {
+		scopedLog.Error(err, "Secrets not found in namespace")
+		return nil, -1, false
+	}
+
+	// secretFound is true when we find atleast one versionedSecretIdentifier based secret AND we can extract its version successfully
+	var secretFound bool = false
+
+	// existingLatestVersion holds the version number of the latest versionedSecretIdentifier based secret, if atleast one exists(defaults to -1)
+	var existingLatestVersion int = -1
+
+	// existingLatestVersionedSecret holds the latest versionedSecretIdentifier based secret, if atleast one exists
+	var existingLatestVersionedSecret corev1.Secret
+
+	// Loop through all secrets in K8S cluster
+	for _, secret := range secretList.Items {
```

Comment:
If only we can sort the secret list based on the version number, then we won't need to loop through all the versioned secrets and get the latest versioned secret. OR if we can only keep the latest version secret only and delete all the previous versions, then we can just get the current latest version and update it.

Reply:
**Deletion of secrets:**
We are handling deletion of secrets separately via a different mechanism. 


**Sorting:**
corev1.SecretList{} doesn't support indexing each entry and thus sorting via a particular field would not be possible without transferring it to a different slice of strings with just names. Since we will take care of deletion of older secrets the operation of running through should be relatively quick. Please let me know your thoughts/alternate approaches to sorting if any.

Note: After the sort, we would need to still search through the sorted list to find the first secret which includes the versionedSecretIdentifier.